docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,287 @@
+
+<!-- Operators can also define how much they tolerate the redundancy to deviate from the desired level. -->
+
+By default, the redundancy is set to `1` and impacts the gossiping of transactions as follows:
+
+
+```go
+  redundancy := duplicateTxs / firstTimeTxs
+  if redundancy < r.redundancyLowerBound:
+    peer.send(Reset)
+  if redundancy > redundancyUpperBound
+    enableRouteDisabling() 
+```
+
+The number of unique and duplicate transactions is upadated as transactions come in and,
+periodically (every `1s`), DOG recomputes the redundnacy of the system. 
+
+
+#### Redundancy adjustment triggering
+
+As explained, redundancy is adjusted periodicially. It is also explicitly triggered when a peer is disconnected. <!-- TODO justify why-->
+
+### Impacted areas of code
+
+The bulk of the changes is constrained to the mempool reactor. 
+
+
+#### Breaking changes and new additions
+
+- The `Mempool` interface is expanded with a method : 
+```go
+// GetSenders returns the list of node IDs from which we receive the given transaction.
+GetSenders(txKey types.TxKey) ([]nodekey.ID, error)
+``` 
+- The `Entry` interface in the mempool package is expanded with a method:
+```go
+// Senders returns the list of registered peers that sent us the transaction.
+	Senders() []nodekey.ID
+```
+- We introduce a new communication channel, called `Mempool Control Channel`. The channel
+is used to transmit the messages needed for the implementation of the protocol. The channel ID
+is `31` and it has a priority of `10` .
+
+Additionally, the mempool reactor is extended with a 
+`gossipRouter` and a `redundancyControl` struct to keep track of the redundancy
+of transactions in the system. 
+Each node keeps track of the routes it disabled between its peers. 
+
+```go
+type gossipRouter struct {
+	mtx cmtsync.RWMutex
+	// A set of `source -> target` routes that are disabled for disseminating transactions. Source
+	// and target are node IDs.
+	disabledRoutes map[nodekey.ID]p2pIDSet
+}
+``` 
+
+
+```go
+type redundancyControl struct {
+	txsPerAdjustment int64
+
+	// Pre-computed upper and lower bounds of accepted redundancy.
+	lowerBound float64
+	upperBound float64
+
+	mtx          cmtsync.RWMutex
+	firstTimeTxs int64 // number of transactions received for the first time
+	duplicates   int64 // number of duplicate transactions
+
+	// If true, do not send HaveTx messages.
+	blockHaveTx atomic.Bool
+}
+
+```
+
+
+#### Added logic to existing functions
+
+The overall behaviour of existing functions remains the same when DOG is disabled. 
+
+Enabling DOG causes the following changes in behaviour:
+
+`Receive`
+
+The function now handles messages from two channels. The `MempoolControlChannel`
+transmits `HaveTx` and `Reset` messages that keep track of routes between peers. 
+
+`TryAddTx` 
+
+When DOG is enabled we count the number of unique and duplicate transactions, and send
+`HaveTx` messages if the redundancy is too high. 
```

Comment:
I thought this action was periodical, not triggered by the reception of a message.

Reply:
The count is done on reception, and HaveTx will be sent only then. Whether it is sent is decided periodically. 

I will clarify this but basically:
- OnNewMessage: 
   incDuplicate
       if(DogEnabled)
          if(HaveTxUnblocked)
             SendHaveTx

another process 

Every 1s {
  redundancy = computeRedundancy()
  if redundancy > threshold
      unblockHaveTx
  if redundancy < threshold
     resetRoutes

}